server/util.go

```diff
@@ -245,75 +245,72 @@ func uniformValue(value interface{}) interface{} {
 	}
 }
 
-func dumpRowValuesBinary(alloc arena.Allocator, columns []*ColumnInfo, row []types.Datum) (data []byte, err error) {
+func dumpRowValuesBinary(alloc arena.Allocator, columns []*ColumnInfo, row []types.Datum, buffer []byte) ([]byte, error) {
 	if len(columns) != len(row) {
-		err = mysql.ErrMalformPacket
-		return
+		return nil, mysql.ErrMalformPacket
 	}
-	data = append(data, mysql.OKHeader)
-	nullsLen := ((len(columns) + 7 + 2) / 8)
-	nulls := make([]byte, nullsLen)
+	buffer[0] = mysql.OKHeader
+	nulls := buffer[1:]
 	for i, val := range row {
 		if val.IsNull() {
 			bytePos := (i + 2) / 8
 			bitPos := byte((i + 2) % 8)
 			nulls[bytePos] |= 1 << bitPos
 		}
 	}
-	data = append(data, nulls...)
 	for i, val := range row {
 		switch val.Kind() {
 		case types.KindInt64:
 			v := val.GetInt64()
 			switch columns[i].Type {
 			case mysql.TypeTiny:
-				data = append(data, byte(v))
+				buffer = append(buffer, byte(v))
 			case mysql.TypeShort, mysql.TypeYear:
-				data = append(data, dumpUint16(uint16(v))...)
+				buffer = append(buffer, dumpUint16(uint16(v))...)
 			case mysql.TypeInt24, mysql.TypeLong:
-				data = append(data, dumpUint32(uint32(v))...)
+				buffer = append(buffer, dumpUint32(uint32(v))...)
 			case mysql.TypeLonglong:
-				data = append(data, dumpUint64(uint64(v))...)
+				buffer = append(buffer, dumpUint64(uint64(v))...)
 			}
 		case types.KindUint64:
 			v := val.GetUint64()
 			switch columns[i].Type {
 			case mysql.TypeTiny:
-				data = append(data, byte(v))
+				buffer = append(buffer, byte(v))
 			case mysql.TypeShort, mysql.TypeYear:
-				data = append(data, dumpUint16(uint16(v))...)
+				buffer = append(buffer, dumpUint16(uint16(v))...)
 			case mysql.TypeInt24, mysql.TypeLong:
-				data = append(data, dumpUint32(uint32(v))...)
+				buffer = append(buffer, dumpUint32(uint32(v))...)
 			case mysql.TypeLonglong:
-				data = append(data, dumpUint64(v)...)
+				buffer = append(buffer, dumpUint64(v)...)
 			}
 		case types.KindFloat32:
 			floatBits := math.Float32bits(val.GetFloat32())
-			data = append(data, dumpUint32(floatBits)...)
+			buffer = append(buffer, dumpUint32(floatBits)...)
 		case types.KindFloat64:
 			floatBits := math.Float64bits(val.GetFloat64())
-			data = append(data, dumpUint64(floatBits)...)
+			buffer = append(buffer, dumpUint64(floatBits)...)
 		case types.KindString, types.KindBytes:
-			data = append(data, dumpLengthEncodedString(val.GetBytes(), alloc)...)
+			buffer = append(buffer, dumpLengthEncodedString(val.GetBytes(), alloc)...)
```

Comment:
`buffer = dumpLengthEncodedString(buffer, val.GetBytes())`
Then we don't need to use `alloc`.